### ISSUE TITLE
Sane defaults cleanup

### ISSUE THREAD

**Issue by pimotte** [syncany/syncany#345]
Currently cleanup fully deletes any deleted files. Versions are kept in a fairly dummy way. These should both be replaced with some solid, sane defaults, such that the correct versions are kept from a user point of view and they get the opportunity to restore deleted versions if necessary.

**Comment by pimotte** [comment]
I started working on this at https://github.com/syncany/syncany/tree/feature/cleanup-sane-defaults

**Comment by binwiederhier** [comment]
Looks perfect so far :+1: 
I don't have *anything* to complain about!!

**Comment by pimotte** [comment]
Okay, so I'm trying to choose which versions to keep. I have the following considerations with regards to the design of this:
- By default all versions should be deleted eventually (unless the file is not changed, then the last version stays).
- Most versions that are kept should be recent. The older versions become, the less should be kept over time.

Nice to have:
- Being able to do the filtering using SQL, so that there is no need to load all the metadata into memory.

I tried to implement this in the following way: 

Have an SQL query that accepts a begin timestamp, an end timestamp and a pattern to format a date. Truncate all timestamps of versions using this pattern, and only keep the most recent one for each of these. 

I have started implementing this here:
https://github.com/syncany/syncany/tree/feature/cleanup-sane-defaults-purging-attempt

What I like about this:
- The heavy lifting can be done in SQL
- It yields a reasonable default.

What I don't like about this.
- It is not that configurable in terms of repo-size. We could start trimming it from the back if it turns out to be too big.
- It is not really configurable in a user friendly way.
- It does not offer fine-grained configuration. (ie. the only options are "keep one per minute/day/week/month/year")

The current default is: Keep all versions from past 5 minutes. Keep one every minute of the past hour. Keep one every hour of the past 3 days. Keep one every day of the past month. Delete everything older than a month.

Before I start patching up tests/testing/code polishing this implementation, I would like some feedback on whether the general concept is something that is sufficient. At first I doubted whether it was good enough, but I'm slowly growing towards this solution, mostly because it is really nice to do all the filtering in the database. Opinions?

**Comment by binwiederhier** [comment]
I'll look at this tonight. From what I read (and understand), it sounds very reasonable. I'll have to re-read it and maybe question you a bit to understand it even more. 

Maybe one remark: I think it's crucial to integrate repo size in the delete considerations. IMHO, people will expect at least the commonly used options "Keep max. X versions" and "Keep max. X day/weeks/...". I think the intelligent delete-logic (keep all in last 5min, ...) might even be a bit much -- dispite the fact that [other projects have a similar logic](http://doc.owncloud.org/server/6.0/user_manual/files/versioncontrol.html).

Let's talk more tonight.

**Comment by pimotte** [comment]
Discussion on IRC lead to:
- Use the magic described above, do not expose options.
- Also have "Delete if older than X days"
- Also have "Delete versions if repo size exceeds X MB"

**Comment by pimotte** [comment]
Status:

Testsuite passes, delete older than X days is implemented. I have chosen to have this be the same option as the time we use for the purging of deleted files. I can't think of any real need to have these differ too much.

The thing that is missing is the third option. The reason is I actually have no idea how to do this in a reasonably smart way, since I don't really have  a clue how to get the size of the repo. The bit that matters is the space occupied by previous versions. However, between it being atrivial to determine how much space is occupied by a version (chunks could be used by other versions) and not repacking multichunks (maybe chunks are in a multichunk that is needed for other files). 
The only thing I can think of is: actually computing the size of the repo, and slowly shortening the time specified by option 2 until we are below the threshold.

Am I missing something here that makes this more doable?

**Comment by binwiederhier** [comment]
**ad "delete if older than X":** So deleted files are now treated the same as the other files: They are deleted after X days? If so: Makes sense!

**ad "repacking multichunks":** I was wondering when this might come up again. We'll eventually have to deal with that as well. I'll make a reminder issue. This sounds to me like a Pim-issue :)

**ad "delete if repo > X MB":** I think your proposed way is the only way to do it -- at least the only way I can think of. I played around with the database a bit, but I think unless we create some unmaintainable monster of a SQL statement, the best option would be what you suggested. 

How about this: Select all file versions and their multichunks (incl. size); order them by database version (newest first). Then we walk from row to row and calculate the rolling sum of the not-yet encountered multichunk sizes. Once we reach the "keep size", we stop and mark the rest for deletion. Too simple? 

Like this:

```sql
select fv.databaseversion_id, fv.filehistory_id, fv.version, mcc.multichunk_id, mc.size
from fileversion fv
join filecontent_chunk fcc on fv.filecontent_checksum=fcc.filecontent_checksum
join multichunk_chunk mcc on fcc.chunk_checksum=mcc.chunk_checksum
join multichunk mc on mcc.multichunk_id=mc.id
group by fv.databaseversion_id, fv.filehistory_id, fv.version, mcc.multichunk_id, mc.size
order by fv.databaseversion_id desc, fv.filehistory_id, fv.version
```

![2015020223 13 15_selection_001](https://cloud.githubusercontent.com/assets/664597/6011288/13d1f188-ab32-11e4-9f6f-b106596a1037.png)

Example (shortened checksums):

```
   DBV HISTORY_ID   VERSION  MULTICHUNK_ID    SIZE (rounded to MB)
0  424 0ca2ede63f59  3       918300feffe30    1.6
1  424 0ec9c53d5751  2       e4cff5fcb1220    4.6
2  424 0ecdb4d89310  2       f196c5f40d36f    4.2
3  424 110fb84de755  3       f196c5f40d36f    4.2
4  424 11c8fc469c82  2       a6130ac316f97    1.0
5  424 120fa6989338  2       282eebaa32e5f    1.4
...
```

Assuming the "keep size" is 12 MB (example!), we would keep the file versions in rows 0..4 (1.6 + 4.6 + 4.2 + 1.0 = 11.4 MB < 12 MB; do not count multichunk f196c5f40d36f twice!), and delete 5..n.

This is a very simple statement and it does not take DELETE versions into account. Nonetheless: Is that similar to what you had in mind?

**Comment by pimotte** [comment]
at repacking: Technically speaking, it's certainly a Pim-issue. Pim-user-needs speaking, not so much:P. I'll probably pick it up at some point though. 

This was more or less what I was gravitating towards. 

Do we want this to be a hard limit? If yes, we don't account for metadata here, so we'd have to take some buffer space. (Extreme case: 300K files of size 512 bytes: 172MiB multichunks, 42MiB database file.)

I also have a gut feeling we could do the entire thing in SQL, but that might be a bit much.

**Comment by binwiederhier** [comment]
I think we should disregard metadata for now. I agree that doing it in SQL will be overkill. Simplicity is king here to make it maintainable. 

What the above logic is obviously also missing is that we need to keep the first occurance of every file version, because that's the "currently active" one. So even after the limit is reached, we need to keep all the "firsts".

**Comment by pimotte** [comment]
Closed by #378
